### PR TITLE
Automation test for checking that google groups are being synced appropriately

### PR DESCRIPTION
**Pull request by marctalbott**
Ticket: [GAWB-3853](https://broadinstitute.atlassian.net/browse/GAWB-3853)
- a test that makes sure that group sync is actually updating the google group correctly

---

**PR checklist**

- [x] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [x] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [x] I've updated the [FISMA documentation](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

### CODE REVIEW

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/SamApiSpec.scala [hunk 1]
```diff
@@ -361,6 +364,52 @@ class SamApiSpec extends FreeSpec with BillingFixtures with Matchers with ScalaF
       // result should be the same
       petEmail2 shouldBe petEmail1
     }
+
+    "should synchronize groups with Google" in {
+      val Seq(user1: Credentials, user2: Credentials, user3: Credentials) = UserPool.chooseStudents(3)
+
+      implicit val ec: ExecutionContextExecutor = ExecutionContext.global
+      val managedGroupId = "fooGroup"
```
Comment (dvoet) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/SamApiSpec.scala]: make this a random name, e.g. a random uuid with a prefix. Using a fixed name like this means concurrent test runs will interfere with each other or a failed cleanup in one test run will cause failure in the next.

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/SamApiSpec.scala [hunk 2]
```diff
@@ -361,6 +364,52 @@ class SamApiSpec extends FreeSpec with BillingFixtures with Matchers with ScalaF
...
+      val Seq(user1Proxy: WorkbenchEmail, user2Proxy: WorkbenchEmail, user3Proxy: WorkbenchEmail) = Seq(user1, user2, user3).map(user => Sam.user.proxyGroup(user.email)(user1AuthToken))
+
+      Sam.user.createGroup(managedGroupId)(user1AuthToken)
+      register cleanUp Sam.user.deleteGroup(managedGroupId)(user1AuthToken)
+
+      Sam.user.setPolicyMembers(managedGroupId, adminPolicy, Set(user1.email, user2.email))(user1AuthToken)
+
+      val policies = Sam.user.listResourcePolicies("managed-group", managedGroupId)(user1AuthToken)
+      val policyEmail = for {
+        policy <- policies.asInstanceOf[Set[Map[String, Map[String, List[String]]]]] if policy.getOrElse("policy", Map.empty).getOrElse("memberEmails", List.empty).nonEmpty
```
Comment (dvoet) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/SamApiSpec.scala]: I don't understand how this casting is working... seems fishy. Why doesn't listResourcePolicies return a structured result?

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/SamApiSpec.scala [hunk 3]
```diff
@@ -361,6 +364,52 @@ class SamApiSpec extends FreeSpec with BillingFixtures with Matchers with ScalaF
...
+      // check that google has users 1 and 2 in it
+      TestKit.awaitCond(
+        Await.result(googleDirectoryDAO.listGroupMembers(WorkbenchEmail(policyEmail.head)), 5.minutes)
+          .getOrElse(Set.empty).toSet == Set(user1Proxy.value, user2Proxy.value),
+        5.minutes, 5.seconds)
+
+      Sam.user.addUserToPolicy(managedGroupId, adminPolicy, user3.email)(user1AuthToken)
+
+      // check that google has users 1, 2, and 3 in it
+      TestKit.awaitCond(
```
Comment (dvoet) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/SamApiSpec.scala]: I found a better way than awaitCond which is awaitAssert (which is in TestKitBase which your test class needs to extend). Inside awaitAssert call your normal asserts.

**Review by dvoet — CHANGES_REQUESTED** [review]

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/SamApiSpec.scala (hunk 1 above)
Comment (Qi77Qi) [reply automation/src/test/scala/org/broadinstitute/dsde/test/api/SamApiSpec.scala]: it's probably worth creating a scalacheck generator for this (similar to https://github.com/broadinstitute/sam/blob/develop/src/test/scala/Generator.scala#L77)

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala [hunk 4]
```diff
@@ -414,6 +414,13 @@ class GoogleExtensions(val directoryDAO: DirectoryDAO, val accessPolicyDAO: Acce
 
         group = groupOption.getOrElse(throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.NotFound, s"$groupId not found")))
 
+        groupOption2 <- groupId match {
+          case basicGroupName: WorkbenchGroupName => directoryDAO.loadGroup(basicGroupName)
```
Comment (Qi77Qi) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala]: think you need to `unsafeToRunSync()` or `.futureValue` on these `loadGroup`?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala [hunk 5]
```diff
@@ -414,6 +414,13 @@ class GoogleExtensions(val directoryDAO: DirectoryDAO, val accessPolicyDAO: Acce
...
+          case rpn: ResourceAndPolicyName => accessPolicyDAO.loadPolicy(rpn)
+        }
+
+        group2 = groupOption2.getOrElse(throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.NotFound, s"$groupId not found")))
```
Comment (Qi77Qi) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala]: you can do `group2 <- IO.fromEither(groupOption2.toRight(throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.NotFound, s"$groupId not found")))).unsafeToFuture` (probably will be easier to move things to IO later this way too)

**Review by dvoet — APPROVED** [review]

**Review by gpolumbo-broad — APPROVED** [review]

**Pull request merged**